Fix/issue 68182: Fix rounding issue in getMaxChildIndexForScrollOffset

daviddgl opened the pull request:
## Description

We're trying to fix a rounding issue describe in here #68182

## Related Issues

Fixes #68182

## Tests

I added the following tests:

Added parameterized tests to check that previous and current behaviour are still working in `sliver_fixed_extent_layout_test.dart` file

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

tsimbalar commented:
@googlebot I fixed it. 

(was missing for my e-mail address)

daviddgl commented:
@googlebot I consent.

daviddgl commented:
@googlebot I consent.

daviddgl commented:
@googlebot I fixed it.

tsimbalar commented:
@googlebot I consent.

daviddgl commented:
Hi @Piinks 
this PR is ready to review, not sure how to remove the `work in progress` label :(
Thanks

Piinks commented on packages/flutter/test/rendering/sliver_fixed_extent_layout_test.dart:
```diff
@@ -41,6 +41,42 @@ void main() {
+
+  group('getMaxChildIndexForScrollOffset', () {
+    const double itemExtent = 600.0;
+    testValidationRounding(1234, 0, 0, 'When item extent is 0');
```
These tests are a bit difficult to understand, can you refactor them into separate tests that use a common method? Or one large test that breaks down each case?

daviddgl replied:
I create a test per case. Thanks

Piinks commented on packages/flutter/test/rendering/sliver_fixed_extent_layout_test.dart:
```diff
@@ -41,6 +41,42 @@ void main() {
+    testValidationRounding(
+        2468.571428571429, 411.42857142857144, 5, 'Edge case for Android');
+    testValidationRounding(
+        2484.0, 414.0, 5, 'Same Android edge case but for iOS');
```
I do not understand how these tests are testing different Android and iOS behavior. How is the platform factored into this?

daviddgl replied:
Actually, it's not because of the platform, it's because of the screen size, the itemExtent property. I'll change the description

Piinks reviewed (commented):
Hi @daviddgl welcome! Thank you for the contribution. 🎉 

It looks like this change may be using dart format. The flutter repository has its own style guide. Can you format to match the guide in the wiki? https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

daviddgl commented:
> Hi @daviddgl welcome! Thank you for the contribution. 🎉
> 
> It looks like this change may be using dart format. The flutter repository has its own style guide. Can you format to match the guide in the wiki? https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

I think I applied the style as the guideline says.
Thanks :)

daviddgl commented:
Hi @Piinks 
Could you re-review this PR, please?

Thanks!

Piinks commented on packages/flutter/test/rendering/sliver_fixed_extent_layout_test.dart:
```diff
@@ -41,6 +41,75 @@ void main() {
+  });
+}
+
+int testGetMaxChildIndexForScrollOffset(
+    double scrollOffset, double itemExtent) {
+  final TestRenderSliverFixedExtentBoxAdaptor sut =
+      TestRenderSliverFixedExtentBoxAdaptor();
```
What does 'sut' stand for? We usually avoid abbreviations.
```suggestion
int testGetMaxChildIndexForScrollOffset(double scrollOffset, double itemExtent) {
  final TestRenderSliverFixedExtentBoxAdaptor sut = TestRenderSliverFixedExtentBoxAdaptor();
```

daviddgl replied:
`sut` means System Under Test
https://en.wikipedia.org/wiki/System_under_test

I'll rename it if you don't use this convention.

Piinks reviewed (commented):
This is looking good @daviddgl. Thanks for the update. Just a few nits, and it looks like the analyzer is failing due to some trailing white space.

daviddgl commented:
@googlebot I fixed it.

daviddgl committed:
fix(itemExtent): Fix rounded issue using precisionErrorTolerance

To calculate the index in getMaxChildIndexForScrollOffset it need to use the precisionErrorTolerance constant to return the correct one.

We add some tests to make sure that the current behaviour it's working before do the implementation.

Co-authored-by: Thibaud Desodt <tibo.desodt@gmail.com>

daviddgl commented:
@googlebot I fixed it.

Piinks reviewed (approved):
This LGTM, although the analyzer is currently failing. Can you check and resolve that trailing white space?

tsimbalar commented on packages/flutter/test/rendering/sliver_fixed_extent_layout_test.dart:
```diff
@@ -103,3 +171,17 @@ class TestRenderSliverBoxChildManager extends RenderSliverBoxChildManager {
 }
+
+class TestRenderSliverFixedExtentBoxAdaptor extends RenderSliverFixedExtentBoxAdaptor {
+  TestRenderSliverFixedExtentBoxAdaptor() 
```
I think the analyzer is complaining about 

```suggestion
  TestRenderSliverFixedExtentBoxAdaptor()
```

tsimbalar replied:
cf https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8865114836733794336/+/steps/prepare_environment/0/steps/Analyze/0/steps/analyze/0/stdout

daviddgl replied:
Thanks, it's fixed.
